common/src/gas_provider/internal.rs

```diff
@@ -629,6 +640,100 @@ where
     }
 
+    // TODO (breathx): doc this.
+    fn lock(key: Self::Key, amount: Self::Balance) -> Result<(), Self::Error> {
+        let node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;
+
+        if node.is_reserved_local() {
+            return Err(InternalError::forbidden().into());
+        }
+
+        if node.is_consumed() {
+            return Err(InternalError::node_was_consumed().into());
+        }
+
+        let (mut upstream_node, node_id) = Self::node_with_value(node)?;
+
+        let upstream_node_value = upstream_node
+            .value_mut()
+            .ok_or_else(InternalError::unexpected_node_type)?;
+
+        if *upstream_node_value < amount {
+            return Err(InternalError::insufficient_balance().into());
+        }
+
+        *upstream_node_value = upstream_node_value.saturating_sub(amount);
+
+        let mut node = if let Some(upstream_id) = node_id {
+            StorageMap::insert(upstream_id, upstream_node);
+
+            Self::get_node(key).ok_or_else(InternalError::node_not_found)?
+        } else {
+            upstream_node
+        };
+
+        let node_lock = node
+            .lock_mut()
+            .ok_or_else(InternalError::unexpected_node_type)?;
+
+        *node_lock = node_lock.saturating_add(amount);
+
+        StorageMap::insert(key, node);
+
+        Ok(())
+    }
+
+    // TODO (breathx): doc this.
+    fn unlock(key: Self::Key, amount: Self::Balance) -> Result<(), Self::Error> {
+        let mut node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;
+
+        if node.is_reserved_local() {
+            return Err(InternalError::forbidden().into());
+        }
+
+        if node.is_consumed() {
+            return Err(InternalError::node_was_consumed().into());
+        }
+
+        let node_lock = node
+            .lock_mut()
+            .ok_or_else(InternalError::unexpected_node_type)?;
+
+        if *node_lock < amount {
+            return Err(InternalError::insufficient_balance().into());
+        }
+
+        *node_lock = node_lock.saturating_sub(amount);
+
+        let (upstream_node, upstream_id) = Self::node_with_value(node.clone())?;
+
+        let (mut ancestor_node, ancestor_id) = if let Some(upstream_id) = upstream_id {
+            StorageMap::insert(key, node);
+
+            (upstream_node, upstream_id)
+        } else {
+            (node, key)
+        };
+
+        let ancestor_value = ancestor_node
+            .value_mut()
+            .ok_or_else(InternalError::unexpected_node_type)?;
+
+        *ancestor_value = ancestor_value.saturating_add(amount);
+
+        StorageMap::insert(ancestor_id, ancestor_node);
+
+        Ok(())
+    }
+
+    fn get_lock(key: Self::Key) -> Result<Self::Balance, Self::Error> {
+        let node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;
+
+        let lock = node.lock().ok_or_else(InternalError::forbidden)?;
+
+        Ok(lock)
```

Comment:
Looks like the returned value can be used directly:
```suggestion
        let node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;

        node.lock().ok_or_else(InternalError::forbidden)
```

Reply:
Yeah, I've tried, but it looks not so good:
```rust
node
    .lock()
    .ok_or_else(|| InternalError::forbidden().into())
```
OR
```rust
node
    .lock()
    .ok_or_else(InternalError::forbidden)
    .map_err(Into::into)
```